### ISSUE TITLE
Doc/README: mention that Spyder needs to be uninstalled before installing a newer version with 'python setup.py install'

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#823]
_From [question...@gmail.com](https://code.google.com/u/111411311807259959395/) on 2011-11-02T13:50:54Z_

Spyder Version:  2.1.1
Python Version:  2.6.5
Qt Version:      4.6.2, PyQt4 (API v1) 4.7.2 on Linux 

What steps will reproduce the problem?

 1. Download, unzip, and install Spyder (`sudo python setup.py install`)
2. Run Spyder (`spyder`) What is the expected output? What do you see instead? Expected: program loads
What actually happens:
Traceback (most recent call last):
  File "/usr/local/bin/spyder", line 2, in \<module>
    from spyderlib import spyder
  File "/usr/local/lib/python2.6/dist-packages/spyderlib/spyder.py", line 99, in \<module>
    from spyderlib.plugins.editor import Editor
  File "/usr/local/lib/python2.6/dist-packages/spyderlib/plugins/editor.py", line 36, in \<module>
    from spyderlib.widgets.editor import (ReadWriteStatus, EncodingStatus,
  File "/usr/local/lib/python2.6/dist-packages/spyderlib/widgets/editor.py", line 38, in \<module>
    from spyderlib.widgets.editortools import OutlineExplorer, check
ImportError: cannot import name OutlineExplorer 

Please provide any additional information below

. I previously had Spyder 2.0.1 installed, which worked fine -- I'm guessing that's what caused the problem somehow?

Also, Spyder 2.1.1 runs fine from source (`python bootstrap.py`) -- it's just not getting installed as a package correctly.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=823_

**Comment by spyder-issues-migrator** [comment]
_From [question...@gmail.com](https://code.google.com/u/111411311807259959395/) on 2011-11-02T11:57:50Z_

I cloned the repository, then did
python setup.py build
sudo python setup.py install

and it works now...

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-11-02T13:16:33Z_

Installing from source package or from the mercurial repository should lead to the same result. 

The traceback suggest that you had some '.pyc' files of Spyder v2.0 that were not removed during uninstall. How did you uninstall Spyder v2.0?

**Comment by spyder-issues-migrator** [comment]
_From [question...@gmail.com](https://code.google.com/u/111411311807259959395/) on 2011-11-02T13:20:01Z_

I didn't uninstall it -- I guess that was the problem! Maybe it should be mentioned in the readme that previous versions should be uninstalled first?

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-11-02T13:33:58Z_

Maybe indeed but this is not mentioned currently because every Python package needs to be uninstalled before installing a newer version (except if you are using installing tools like easy_install: pip, setuptools, ...). 

Anyway, I will mention this in the readme file.

**Summary:** Doc/README: mention that Spyder needs to be uninstalled before installing a newer version with 'python setup.py install'  
  **Status:** Accepted  
  **Labels:** -Type-Defect Type-Task Milestone-Release2.1

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-11-02T23:16:46Z_

I am not sure if we should support installation with `python setup.py install` at all if it causes such problems.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-11-03T00:06:38Z_

No I think that it's not a problem because people installing Spyder this way will either know what they are doing (they will know that uninstall must be done first without having to read the readme file) or they will choose this method of installation after reading the doc/readme so they will know what to do if we mention it. 

Besides it will be a good thing for novice users to know when they install other Python packages.

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-11-03T01:03:30Z_

But `python setup.py` doesn't support uninstallation. I believe it is only supported by binary .exe installers. So you need to describe in detail how `uninstall` works.

**Comment by spyder-issues-migrator** [comment]
_From [question...@gmail.com](https://code.google.com/u/111411311807259959395/) on 2011-11-03T01:14:58Z_

I'm by no means an expert in Python, but nonetheless, I've never had a problem installing a new version of a package over an older one, so I was surprised to hear that "uninstall must be done first". On most of the forums I found where people were complaining about the difficulty of uninstalling Python packages, the most common response was "Why bother?". For example: http://www.velocityreviews.com/forums/t338779-why-no-python-setup-py-uninstall.html The consensus seems to be that if Python packages needed to be uninstalled regularly, there would be an easier way to do it.

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-11-03T01:32:36Z_

The problem is in our plugin discovery. In standard Python packages files are usually not removed from release to release, and even if they are removed, new code doesn't try to import them. Spyder loads all plugins it can find, so if there are pieces from the old version - it will try to load them as well.

We can hardcode a list of standard plugins to load from Spyder location, but that would be a maintenance overhead. I am not sure that a right solution is. Perhaps fill a bug to distutils.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-11-03T02:29:12Z_

@-questionably: most of the time, uninstalling is not strictly necessary (it simply works without it) but you may have trouble from time to time if you simply install packages over packages. Anyway, it's far cleaner to uninstall any Python package before installing a new version with `python setup.py install`. The fact that it works mostly without uninstallation does not mean that uninstallation is not needed. Anyway, the uninstallation support is one of the motivation that lead to the development of setuptools, then distribute and pip. The fact is that if you don't uninstall a Python package, if the next version is a major evolution of the code (say that a whole subpackage was removed or renamed) then you will have tons of files that are no longer used by your system. There is no need to be an expert in Python to understand that -after a lot of installations- this will result in a big mess in your Python site-packages directory.

@-techtonik: if we fill a but to distutils, I know the answer that we would have: use `pip`, as recommended on PyPI ( http://pypi.python.org/pypi ).

Anyway, the problem here may be different as the traceback suggests that "spyderlib/widgets/editor.py" file was still v2.0 which is weird as this file exists in both versions. (@techtonik: the plugin discovery has nothing to do here because this process is just used for extra plugins, i.e. modules of the spyderplugins package)
Moreover the fact that it failed with your first attempt from the source package and suddenly succeeds with the same procedure from the source repository is very suspiscious.

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-11-03T02:39:55Z_

I've just noticed that the problem is with Linux. From my experience with Python in Linux it is better to stick to the packages provided by the platform. In case of Spyder - I execute it from source.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-11-03T02:44:28Z_

Using your Linux package manager is the wisest solution indeed.

**Comment by spyder-issues-migrator** [comment]
_From [steve.f....@gmail.com](https://code.google.com/u/117650558843796505335/) on 2011-11-04T14:45:51Z_

So what's the end solution here?

I've now read the docs and see that there's a bootstrap script that is the preferred method of running from source.  In a perfect world everyone reads all the docs all the time and never has problems.  The thing I don't like about this process is there are multiple ways to run where one is common to lots of python applications ("python setup.py install") and one is not ("custom" method - bootstrap), but the common one doesn't work seamlessly.  Maybe the bootstrap method is common but I've just never ran into it before?  Having a common method that doesn't always just work seems like a bad idea.  I'd say improve it or remove it altogether.

I've had issues getting beta/alpha versions to work due to import issues and it's always made me regret trying to use those versions.  Perhaps this was the problem all along?  If you want more beta testers, this should work as seamlessly as possible.

Feel free to send to /dev/null.  :)

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-11-15T08:05:34Z_

This issue was closed by revision bb5ebafcddf1 .

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [breed...@gmail.com](https://code.google.com/u/105348566576685241380/) on 2014-09-12T09:48:43Z_

Thanks for adding the note to the read me.  Fully uninstalling spyder solved my problem.

I can't get pip to work due to proxy issues, so python setup.py install is the way I have to install spyder on Linux.  I first installed spyder, but I had some older versions of Qt so it wouldn't launch.  I then updated Qt, and tried reinstalling spyder.  I got a new error, but it still would not launch.  Once I uninstalled spyder and reinstalled, then spyder launched.
